

# Learning Bayesian Networks


Previous notebooks showed how Bayesian networks economically encode a probability distribution over a set of variables, and how they can be used e.g. to predict variable states, or to generate new samples from the joint distribution. This section will be about obtaining a Bayesian network, given a set of sample data. Learning a Bayesian network can be split into two problems:

 **Parameter learning:** Given a set of data samples and a DAG that captures the dependencies between the variables, estimate the (conditional) probability distributions of the individual variables.
 
 **Structure learning:** Given a set of data samples, estimate a DAG that captures the dependencies between the variables.
 
This notebook aims to illustrate how parameter learning and structure learning can be done with pgmpy.
Currently, the library supports:
 - Parameter learning for *discrete* nodes:
   - Maximum Likelihood Estimation
   - Bayesian Estimation
 - Structure learning for *discrete*, *fully observed* networks:
   - Score-based structure estimation (BIC/BDeu/K2 score; exhaustive search, hill climb/tabu search)
   - Constraint-based structure estimation (PC)
   - Hybrid structure estimation (MMHC)


## Parameter Learning

Suppose we have the following data:

In [1]:
import pandas as pd
data = pd.DataFrame(data={'fruit': ["banana", "apple", "banana", "apple", "banana","apple", "banana", 
                                    "apple", "apple", "apple", "banana", "banana", "apple", "banana",], 
                          'tasty': ["yes", "no", "yes", "yes", "yes", "yes", "yes", 
                                    "yes", "yes", "yes", "yes", "no", "no", "no"], 
                          'size': ["large", "large", "large", "small", "large", "large", "large",
                                    "small", "large", "large", "large", "large", "small", "small"]})
print(data)

     fruit tasty   size
0   banana   yes  large
1    apple    no  large
2   banana   yes  large
3    apple   yes  small
4   banana   yes  large
5    apple   yes  large
6   banana   yes  large
7    apple   yes  small
8    apple   yes  large
9    apple   yes  large
10  banana   yes  large
11  banana    no  large
12   apple    no  small
13  banana    no  small


In [29]:
data_df = pd.read_csv('superheroes_power_matrix.csv')
list(data_df.columns)

['Name',
 'Agility',
 'Accelerated Healing',
 'Lantern Power Ring',
 'Dimensional Awareness',
 'Cold Resistance',
 'Durability',
 'Stealth',
 'Energy Absorption',
 'Flight',
 'Danger Sense',
 'Underwater breathing',
 'Marksmanship',
 'Weapons Master',
 'Power Augmentation',
 'Animal Attributes',
 'Longevity',
 'Intelligence',
 'Super Strength',
 'Cryokinesis',
 'Telepathy',
 'Energy Armor',
 'Energy Blasts',
 'Duplication',
 'Size Changing',
 'Density Control',
 'Stamina',
 'Astral Travel',
 'Audio Control',
 'Dexterity',
 'Omnitrix',
 'Super Speed',
 'Possession',
 'Animal Oriented Powers',
 'Weapon-based Powers',
 'Electrokinesis',
 'Darkforce Manipulation',
 'Death Touch',
 'Teleportation',
 'Enhanced Senses',
 'Telekinesis',
 'Energy Beams',
 'Magic',
 'Hyperkinesis',
 'Jump',
 'Clairvoyance',
 'Dimensional Travel',
 'Power Sense',
 'Shapeshifting',
 'Peak Human Condition',
 'Immortality',
 'Camouflage',
 'Element Control',
 'Phasing',
 'Astral Projection',
 'Electrical Transport',

We know that the variables relate as follows:

In [2]:
from pgmpy.models import BayesianModel

model = BayesianModel([('fruit', 'tasty'), ('size', 'tasty')])  # fruit -> tasty <- size

Parameter learning is the task to estimate the values of the conditional probability distributions (CPDs), for the variables `fruit`, `size`, and `tasty`. 

#### State counts
To make sense of the given data, we can start by counting how often each state of the variable occurs. If the variable is dependent on parents, the counts are done conditionally on the parents states, i.e. for seperately for each parent configuration:

In [3]:
from pgmpy.estimators import ParameterEstimator
pe = ParameterEstimator(model, data)
print("\n", pe.state_counts('fruit'))  # unconditional
print("\n", pe.state_counts('tasty'))  # conditional on fruit and size


         fruit
apple       7
banana      7

 fruit apple       banana      
size  large small  large small
tasty                         
no      1.0   1.0    1.0   1.0
yes     3.0   2.0    5.0   0.0


We can see, for example, that as many apples as bananas were observed and that `5` large bananas were tasty, while only `1` was not.

#### Maximum Likelihood Estimation

A natural estimate for the CPDs is to simply use the *relative frequencies*, with which the variable states have occured. We observed `7 apples` among a total of `14 fruits`, so we might guess that about `50%` of `fruits` are `apples`.

This approach is *Maximum Likelihood Estimation (MLE)*. According to MLE, we should fill the CPDs in such a way, that $P(\text{data}|\text{model})$ is maximal. This is achieved when using the *relative frequencies*. See [1], section 17.1 for an introduction to ML parameter estimation. pgmpy supports MLE as follows:

In [4]:
from pgmpy.estimators import MaximumLikelihoodEstimator
mle = MaximumLikelihoodEstimator(model, data)
print(mle.estimate_cpd('fruit'))  # unconditional
print(mle.estimate_cpd('tasty'))  # conditional

╒═══════════════╤═════╕
│ fruit(apple)  │ 0.5 │
├───────────────┼─────┤
│ fruit(banana) │ 0.5 │
╘═══════════════╧═════╛
╒════════════╤══════════════╤════════════════════╤═════════════════════╤═══════════════╕
│ fruit      │ fruit(apple) │ fruit(apple)       │ fruit(banana)       │ fruit(banana) │
├────────────┼──────────────┼────────────────────┼─────────────────────┼───────────────┤
│ size       │ size(large)  │ size(small)        │ size(large)         │ size(small)   │
├────────────┼──────────────┼────────────────────┼─────────────────────┼───────────────┤
│ tasty(no)  │ 0.25         │ 0.3333333333333333 │ 0.16666666666666666 │ 1.0           │
├────────────┼──────────────┼────────────────────┼─────────────────────┼───────────────┤
│ tasty(yes) │ 0.75         │ 0.6666666666666666 │ 0.8333333333333334  │ 0.0           │
╘════════════╧══════════════╧════════════════════╧═════════════════════╧═══════════════╛


`mle.estimate_cpd(variable)` computes the state counts and divides each cell by the (conditional) sample size. The `mle.get_parameters()`-method returns a list of CPDs for all variable of the model.

The built-in `fit()`-method of `BayesianModel` provides more convenient access to parameter estimators:


In [5]:
# Calibrate all CPDs of `model` using MLE:
model.fit(data, estimator=MaximumLikelihoodEstimator)


While very straightforward, the ML estimator has the problem of *overfitting* to the data. In above CPD, the probability of a large banana being tasty is estimated at `0.833`, because `5` out of `6` observed large bananas were tasty. Fine. But note that the probability of a small banana being tasty is estimated at `0.0`, because we  observed only one small banana and it happened to be not tasty. But that should hardly make us certain that small bananas aren't tasty!
We simply do not have enough observations to rely on the observed frequencies. If the observed data is not representative for the underlying distribution, ML estimations will be extremly far off. 

When estimating parameters for Bayesian networks, lack of data is a frequent problem. Even if the total sample size is very large, the fact that state counts are done conditionally for each parents configuration causes immense fragmentation. If a variable has 3 parents that can each take 10 states, then state counts will be done seperately for `10^3 = 1000` parents configurations. This makes MLE very fragile and unstable for learning Bayesian Network parameters. A way to mitigate MLE's overfitting is *Bayesian Parameter Estimation*.

#### Bayesian Parameter Estimation

The Bayesian Parameter Estimator starts with already existing prior CPDs, that express our beliefs about the variables *before* the data was observed. Those "priors" are then updated, using the state counts from the observed data. See [1], Section 17.3 for a general introduction to Bayesian estimators.

One can think of the priors as consisting in *pseudo state counts*, that are added to the actual counts before normalization.
Unless one wants to encode specific beliefs about the distributions of the variables, one commonly chooses uniform priors, i.e. ones that deem all states equiprobable.

A very simple prior is the so-called *K2* prior, which simply adds `1` to the count of every single state.
A somewhat more sensible choice of prior is *BDeu* (Bayesian Dirichlet equivalent uniform prior). For BDeu we need to specify an *equivalent sample size* `N` and then the pseudo-counts are the equivalent of having observed `N` uniform samples of each variable (and each parent configuration). In pgmpy:




In [6]:
from pgmpy.estimators import BayesianEstimator
est = BayesianEstimator(model, data)

print(est.estimate_cpd('tasty', prior_type='BDeu', equivalent_sample_size=10))

╒════════════╤═════════════════════╤════════════════════╤════════════════════╤═════════════════════╕
│ fruit      │ fruit(apple)        │ fruit(apple)       │ fruit(banana)      │ fruit(banana)       │
├────────────┼─────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ size       │ size(large)         │ size(small)        │ size(large)        │ size(small)         │
├────────────┼─────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ tasty(no)  │ 0.34615384615384615 │ 0.4090909090909091 │ 0.2647058823529412 │ 0.6428571428571429  │
├────────────┼─────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ tasty(yes) │ 0.6538461538461539  │ 0.5909090909090909 │ 0.7352941176470589 │ 0.35714285714285715 │
╘════════════╧═════════════════════╧════════════════════╧════════════════════╧═════════════════════╛


The estimated values in the CPDs are now more conservative. In particular, the estimate for a small banana being not tasty is now around `0.64` rather than `1.0`. Setting `equivalent_sample_size` to `10` means that for each parent configuration, we add the equivalent of 10 uniform samples (here: `+5` small bananas that are tasty and `+5` that aren't).

`BayesianEstimator`, too, can be used via the `fit()`-method. Full example:

In [7]:
import numpy as np
import pandas as pd
from pgmpy.models import BayesianModel
from pgmpy.estimators import BayesianEstimator

# generate data
data = pd.DataFrame(np.random.randint(low=0, high=2, size=(5000, 4)), columns=['A', 'B', 'C', 'D'])
model = BayesianModel([('A', 'B'), ('A', 'C'), ('D', 'C'), ('B', 'D')])

model.fit(data, estimator=BayesianEstimator, prior_type="BDeu") # default equivalent_sample_size=5
for cpd in model.get_cpds():
    print(cpd)


╒══════╤══════════╕
│ A(0) │ 0.515185 │
├──────┼──────────┤
│ A(1) │ 0.484815 │
╘══════╧══════════╛
╒══════╤════════════════════╤═════════════════════╕
│ B    │ B(0)               │ B(1)                │
├──────┼────────────────────┼─────────────────────┤
│ D(0) │ 0.5060471679096956 │ 0.49247375717963954 │
├──────┼────────────────────┼─────────────────────┤
│ D(1) │ 0.4939528320903044 │ 0.5075262428203605  │
╘══════╧════════════════════╧═════════════════════╛
╒══════╤═════════════════════╤═════════════════════╤═══════════════════╤═════════════════════╕
│ A    │ A(0)                │ A(0)                │ A(1)              │ A(1)                │
├──────┼─────────────────────┼─────────────────────┼───────────────────┼─────────────────────┤
│ D    │ D(0)                │ D(1)                │ D(0)              │ D(1)                │
├──────┼─────────────────────┼─────────────────────┼───────────────────┼─────────────────────┤
│ C(0) │ 0.49594567200486517 │ 0.49330979371863964 │ 0.497628

## Structure Learning

To learn model structure (a DAG) from a data set, there are two broad techniques:

 - score-based structure learning
 - constraint-based structure learning

The combination of both techniques allows further improvement:
 - hybrid structure learning

We briefly discuss all approaches and give examples.

### Score-based Structure Learning


This approach construes model selection as an optimization task. It has two building blocks:

- A _scoring function_ $s_D\colon M \to \mathbb R$ that maps models to a numerical score, based on how well they fit to a given data set $D$.
- A _search strategy_ to traverse the search space of possible models $M$ and select a model with optimal score.


#### Scoring functions

Commonly used scores to measure the fit between model and data are _Bayesian Dirichlet scores_ such as *BDeu* or *K2* and the _Bayesian Information Criterion_ (BIC, also called MDL). See [1], Section 18.3 for a detailed introduction on scores. As before, BDeu is dependent on an equivalent sample size.

In [8]:
import pandas as pd
import numpy as np
from pgmpy.estimators import BdeuScore, K2Score, BicScore
from pgmpy.models import BayesianModel

# create random data sample with 3 variables, where Z is dependent on X, Y:
#data = pd.DataFrame(np.random.randint(0, 4, size=(5000, 2)), columns=list('XY'))
#data['Z'] = data['X'] + data['Y']

bdeu = BdeuScore(data_df, equivalent_sample_size=5)
k2 = K2Score(data)
bic = BicScore(data)

model1 = BayesianModel([('X', 'Z'), ('Y', 'Z')])  # X -> Z <- Y
model2 = BayesianModel([('X', 'Z'), ('X', 'Y')])  # Y <- X -> Z


print(bdeu.score(model1))
print(k2.score(model1))
print(bic.score(model1))

print(bdeu.score(model2))
print(k2.score(model2))
print(bic.score(model2))


-13936.101051153362
-14326.88012027081
-14292.1400887
-20902.744280734016
-20929.567083476162
-20946.7926535


While the scores vary slightly, we can see that the correct `model1` has a much higher score than `model2`.
Importantly, these scores _decompose_, i.e. they can be computed locally for each of the variables given their potential parents, independent of other parts of the network:

In [9]:
print(bdeu.local_score('Z', parents=[]))
print(bdeu.local_score('Z', parents=['X']))
print(bdeu.local_score('Z', parents=['X', 'Y']))

-9195.779551068787
-6993.305836458767
-57.11877777069253


#### Search strategies
The search space of DAGs is super-exponential in the number of variables and the above scoring functions allow for local maxima. The first property makes exhaustive search intractable for all but very small networks, the second prohibits efficient local optimization algorithms to always find the optimal structure. Thus, identifiying the ideal structure is often not tractable. Despite these bad news, heuristic search strategies often yields good results.

If only few nodes are involved (read: less than 5), `ExhaustiveSearch` can be used to compute the score for every DAG and returns the best-scoring one:

In [10]:
from pgmpy.estimators import ExhaustiveSearch

es = ExhaustiveSearch(data, scoring_method=bic)
best_model = es.estimate()
print(best_model.edges())

print("\nAll DAGs by score:")
for score, dag in reversed(es.all_scores()):
    print(score, dag.edges())

[('Y', 'Z'), ('X', 'Z')]

All DAGs by score:
-14292.1400887 [('Y', 'Z'), ('X', 'Z')]
-14326.6985561 [('X', 'Y'), ('X', 'Z'), ('Z', 'Y')]
-14326.6985561 [('Y', 'Z'), ('X', 'Y'), ('X', 'Z')]
-14326.6985561 [('Y', 'X'), ('Y', 'Z'), ('Z', 'X')]
-14326.6985561 [('Y', 'X'), ('Y', 'Z'), ('X', 'Z')]
-14326.6985561 [('Y', 'X'), ('Z', 'Y'), ('Z', 'X')]
-14326.6985561 [('X', 'Y'), ('Z', 'Y'), ('Z', 'X')]
-16497.4392874 [('X', 'Y'), ('Z', 'Y')]
-16500.63996 [('Y', 'X'), ('Z', 'X')]
-18738.2927822 [('Y', 'Z'), ('Z', 'X')]
-18738.2927822 [('X', 'Z'), ('Z', 'Y')]
-18738.2927822 [('Z', 'Y'), ('Z', 'X')]
-20909.0335135 [('Y', 'Z')]
-20909.0335135 [('Z', 'Y')]
-20912.2341861 [('X', 'Z')]
-20912.2341861 [('Z', 'X')]
-20943.5919809 [('Y', 'Z'), ('X', 'Y')]
-20943.5919809 [('Y', 'X'), ('Z', 'Y')]
-20943.5919809 [('Y', 'X'), ('Y', 'Z')]
-20946.7926535 [('Y', 'X'), ('X', 'Z')]
-20946.7926535 [('X', 'Y'), ('X', 'Z')]
-20946.7926535 [('X', 'Y'), ('Z', 'X')]
-23082.9749174 []
-23117.5333848 [('Y', 'X')]
-23117.

Once more nodes are involved, one needs to switch to heuristic search. `HillClimbSearch` implements a greedy local search that starts from the DAG `start` (default: disconnected DAG) and proceeds by iteratively performing single-edge manipulations that maximally increase the score. The search terminates once a local maximum is found.




In [9]:
from pgmpy.estimators import HillClimbSearch
import numpy as np
import pandas as pd
from pgmpy.estimators import BdeuScore, K2Score, BicScore
from pgmpy.models import BayesianModel

data_avengors = pd.read_csv("Avengers_superhero_powers.csv")

# create some data with dependencies


data_df = data_avengors[["Agility","Intelligence","Mind Control","Camouflage","Immortality"]]

hc = HillClimbSearch(data_df, scoring_method = BicScore(data_df))
best_model = hc.estimate()
best_model.edges()

[('Immortality', 'Mind Control')]

In [36]:
sum(data_avengors["Intelligence"])#False are all false...no any true so they can be removed

7

In [38]:
cols = []
for col in data_avengors.columns:
    if sum(data_avengors[col])>4:
        cols.append(col)
len(cols)

TypeError: unsupported operand type(s) for +: 'int' and 'str'

In [25]:
data_avengors_reduced = data_avengors[cols]
data_avengors_reduced

Unnamed: 0  minutes  seconds  Agility  Accelerated Healing  \
0            0        3       30     True                False   
1            1       47       45     True                False   
2            2       49       60     True                False   
3            3        2       90    False                False   
4            4       27       45     True                False   
5            5       75       60    False                 True   
6            6       10        0    False                False   
7            7       21        0     True                False   
8            8        7       15     True                 True   
9            9       10       15     True                False   
10          10        4       75    False                False   
11          11       20        0     True                 True   
12          12       24       75    False                 True   
13          13        5       45    False                 True   
14          14        3       15    False                False   
15          15        4        0    False                False   
16          16       31       30     True                 True   
17          17       46       30     True                 True   
18          18       17        0     True                False   
19          19       16       15    False                False   

    Dimensional Awareness  Cold Resistance  Durability  Stealth  \
0                   False            False       False     True   
1                   False            False       False     True   
2                   False            False        True     True   
3                   False            False       False    False   
4                   False            False       False     True   
5                   False            False        True    False   
6                   False            False        True    False   
7                    True            False       False    False   
8                   False            False        True     True   
9                   False            False        True    False   
10                  False            False        True    False   
11                  False            False        True     True   
12                  False            False        True    False   
13                  False            False        True    False   
14                  False            False        True    False   
15                  False            False       False    False   
16                  False             True        True    False   
17                  False            False        True    False   
18                  False            False        True     True   
19                  False            False        True     True   

    Energy Absorption  ...  Wallcrawling  Vision - Night  Matter Absorption  \
0               False  ...         False            True              False   
1               False  ...         False           False              False   
2               False  ...         False           False              False   
3               False  ...         False           False              False   
4               False  ...         False           False              False   
5                True  ...         False            True              False   
6               False  ...         False           False              False   
7               False  ...         False           False              False   
8               False  ...          True            True              False   
9               False  ...         False           False              False   
10               True  ...         False            True              False   
11              False  ...         False           False              False   
12              False  ...         False           False              False   
13              False  ...         False           False              False   
14              False  .

In [27]:
hc = HillClimbSearch(data_avengors_reduced, scoring_method = BicScore(data_avengors_reduced))
best_model = hc.estimate()
best_model.edges()

KeyboardInterrupt: 

In [19]:
len(data_avengors.columns)
171-89

82

In [23]:
data_avengors.dtypes
death_list = ["Black Panther","Spider-Man","Doctor Strange","Winter Soldier","Scarlet Witch",
                  "Maria Hill","Nick Fury","Gamora","Loki","Hawkeye","Vision"]


Tony Stark/Iron Man
Steve Rogers/Captain America
Thor
Bruce Banner/Hulk
Natasha Romanov/Black Widow
James Rhodes/War Machine
Rocket Raccoon
Nebula
Wong
Shuri
Okoye
M’Baku
Aunt May Parker
Howard the Duck
Captain Marvel
Scott Lang/Ant-Man (trapped in the Quantum Realm)

The search correctly identifies e.g. that `B` and `C` do not influnce `H` directly, only through `A` and of course that `D`, `E`, `F` are independent.


To enforce a wider exploration of the search space, the search can be enhanced with a tabu list. The list keeps track of the last `n` modfications; those are then not allowed to be reversed, regardless of the score. Additionally a `white_list` or `black_list` can be supplied to restrict the search to a particular subset or to exclude certain edges. The parameter `max_indegree` allows to restrict the maximum number of parents for each node.


### Constraint-based Structure Learning

A different, but quite straightforward approach to build a DAG from data is this:

1. Identify independencies in the data set using hypothesis tests 
2. Construct DAG (pattern) according to identified independencies

#### (Conditional) Independence Tests

Independencies in the data can be identified using chi2 conditional independence tests. To this end, constraint-based estimators in pgmpy have a `test_conditional_independence(X, Y, Zs)`-method, that performs a hypothesis test on the data sample. It allows to check if `X` is independent from `Y` given a set of variables `Zs`:

In [12]:
from pgmpy.estimators import ConstraintBasedEstimator

data = pd.DataFrame(np.random.randint(0, 3, size=(2500, 8)), columns=list('ABCDEFGH'))
data['A'] += data['B'] + data['C']
data['H'] = data['G'] - data['A']
data['E'] *= data['F']

est = ConstraintBasedEstimator(data)

print(est.test_conditional_independence('B', 'H'))          # dependent
print(est.test_conditional_independence('B', 'E'))          # independent
print(est.test_conditional_independence('B', 'H', ['A']))   # independent
print(est.test_conditional_independence('A', 'G'))          # independent
print(est.test_conditional_independence('A', 'G',  ['H']))  # dependent

(642.19630389260817, 9.2182479447645071e-119, True)
(7.5715895908583608, 0.75108812998781649, True)
(10.491889195302004, 0.99999995850445556, True)
(8.3895215910208059, 0.98896079442083529, True)
(4614.0, 0.0, True)


`test_conditional_independence()` returns a tripel `(chi2, p_value, sufficient_data)`, consisting in the computed chi2 test statistic, the `p_value` of the test, and a heuristig flag that indicates if the sample size was sufficient. The `p_value` is the probability of observing the computed chi2 statistic (or an even higher chi2 value), given the null hypothesis that X and Y are independent given Zs.

This can be used to make independence judgements, at a given level of significance:

In [13]:
def is_independent(X, Y, Zs=[], significance_level=0.05):
    return est.test_conditional_independence(X, Y, Zs)[1] >= significance_level

print(is_independent('B', 'H'))
print(is_independent('B', 'E'))
print(is_independent('B', 'H', ['A']))
print(is_independent('A', 'G'))
print(is_independent('A', 'G', ['H']))

False
True
True
True
False


#### DAG (pattern) construction

With a method for independence testing at hand, we can construct a DAG from the data set in three steps:
1. Construct an undirected skeleton - `estimate_skeleton()`
2. Orient compelled edges to obtain partially directed acyclid graph (PDAG; I-equivalence class of DAGs) - `skeleton_to_pdag()`
3. Extend DAG pattern to a DAG by conservatively orienting the remaining edges in some way - `pdag_to_dag()`

Step 1.&2. form the so-called PC algorithm, see [2], page 550. PDAGs are `DirectedGraph`s, that may contain both-way edges, to indicate that the orientation for the edge is not determined.




In [14]:
skel, seperating_sets = est.estimate_skeleton(significance_level=0.01)
print("Undirected edges: ", skel.edges())

pdag = est.skeleton_to_pdag(skel, seperating_sets)
print("PDAG edges:       ", pdag.edges())

model = est.pdag_to_dag(pdag)
print("DAG edges:        ", model.edges())

Undirected edges:  [('A', 'C'), ('A', 'H'), ('A', 'B'), ('E', 'F'), ('G', 'H')]
PDAG edges:        [('F', 'E'), ('A', 'H'), ('C', 'A'), ('G', 'H'), ('E', 'F'), ('B', 'A')]
DAG edges:         [('F', 'E'), ('A', 'H'), ('C', 'A'), ('G', 'H'), ('B', 'A')]


The `estimate()`-method provides a shorthand for the three steps above and directly returns a `BayesianModel`:



In [15]:
print(est.estimate(significance_level=0.01).edges())

[('F', 'E'), ('A', 'H'), ('C', 'A'), ('G', 'H'), ('B', 'A')]


The `estimate_from_independencies()`-method can be used to construct a `BayesianModel` from a provided *set of independencies* (see class documentation for further features & methods):

In [16]:
from pgmpy.independencies import Independencies

ind = Independencies(['B', 'C'],
                     ['A', ['B', 'C'], 'D'])
ind = ind.closure()  # required (!) for faithfulness

model = ConstraintBasedEstimator.estimate_from_independencies("ABCD", ind)

print(model.edges())

[('A', 'D'), ('B', 'D'), ('C', 'D')]


PC PDAG construction is only guaranteed to work under the assumption that the identified set of independencies is *faithful*, i.e. there exists a DAG that exactly corresponds to it. Spurious dependencies in the data set can cause the reported independencies to violate faithfulness. It can happen that the estimated PDAG does not have any faithful completions (i.e. edge orientations that do not introduce new v-structures). In that case a warning is issued.


### Hybrid Structure Learning

The MMHC algorithm [3] combines the constraint-based and score-based method. It has two parts:

1. Learn undirected graph skeleton using the constraint-based construction procedure MMPC
2. Orient edges using score-based optimization (BDeu score + modified hill-climbing)

We can perform the two steps seperately, more or less as follows:

In [4]:
from pgmpy.estimators import MmhcEstimator

data = pd.DataFrame(np.random.randint(0, 3, size=(2500, 8)), columns=list('ABCDEFGH'))
data['A'] += data['B'] + data['C']
data['H'] = data['G'] - data['A']
data['E'] *= data['F']

mmhc = MmhcEstimator(data)
skeleton = mmhc.mmpc()
print("Part 1) Skeleton: ", skeleton.edges())

# use hill climb search to orient the edges:
hc = HillClimbSearch(data, scoring_method=BDeuScore(data))
model = hc.estimate(tabu=10, white_list=skeleton.to_directed().edges())
print("Part 2) Model:    ", model.edges())

AttributeError: 'StructureEstimator' object has no attribute 'mmpc'

`MmhcEstimator.estimate()` is a shorthand for both steps and directly estimates a `BayesianModel`.

### Conclusion

This notebook aimed to give an overview of pgmpy's estimators for learning Bayesian network structure and parameters. For more information about the individual functions see their docstring documentation. If you used pgmpy's structure learning features to satisfactorily learn a non-trivial network from real data, feel free to drop us an eMail via the mailing list or just open a Github issue. We'd like to put your network in the examples-section!

### References

[1] Koller & Friedman, Probabilistic Graphical Models - Principles and Techniques, 2009

[2] Neapolitan, [Learning Bayesian Networks](http://www.cs.technion.ac.il/~dang/books/Learning%20Bayesian%20Networks&#40;Neapolitan,%20Richard&#41;.pdf), 2003

[3] Tsamardinos et al., [The max-min hill-climbing BN structure learning algorithm](http://www.dsl-lab.org/supplements/mmhc_paper/paper_online.pdf), 2005



To do :
 1) Reduce the columns , remove columns with all false
 2) Most important powers